Ordre aléatoire des têtes du Royaume

Steellgold opened the pull request:
J'ai remarqué que les photos de profiles était tout le temps aligné dans le même ordre et je trouve sa pas très esthétique

J'ai donc fait en sorte qu'elles soit mélangé pour ajouter un peu plus de réalisme 

Avant vs Après
![image](https://github.com/Virtual-Royaume/Royaume-Frontend/assets/51505384/bd34556b-332b-4497-97e1-25dea54f5dd4)
![image](https://github.com/Virtual-Royaume/Royaume-Frontend/assets/51505384/c67e7a8a-9a8c-46fa-9791-2e2537d61432)

Capture d'écrans prise au moment du rechargement de page

Steellgold committed:
feat(members-galery.tsx): add shuffle function to shuffle members array before rendering them in the gallery

Bluzzi reviewed (changes requested):
L'objectif de ce genre de données c'est juste d'avoir une preview du résultat final sans ajouter de complexité au code.

Si on en ajoute trop, le refactor sera + complexe quand on va connecter le site à l'API.

Steellgold commented:
> L'objectif de ce genre de données c'est juste d'avoir une preview du résultat final sans ajouter de complexité au code.
> 
> Si on en ajoute trop, le refactor sera + complexe quand on va connecter le site à l'API.

Je pense pas que cela rend + difficile au passage à l'API c'est simplement un tableau qu'on donne à la fonction donc il sera du même type qu'actuellement

Bluzzi commented:
Ça peut porter à confusion pour le dev qui s’en occupera, il ne sera pas quoi faire de cette fonction.

Pareil si on crée ce genre de fonction, il faut l’accompagner de test unitaire. Aucune fonction de ce type ne doit pas en avoir.

Steellgold commented:
> Ça peut porter à confusion pour le dev qui s’en occupera, il ne sera pas quoi faire de cette fonction.
> 
> Pareil si on crée ce genre de fonction, il faut l’accompagner de test unitaire. Aucune fonction de ce type ne doit pas en avoir.

@Bluzzi Si je fait un test unitaire et que la fonction je la met donc à part dans un dossier (surement /utils) cela peut aider ?

Bluzzi commented:
C'est déjà + ok, ça m'embête de prendre du temps sur un truc pas forcément permanent, mais oui tu peux.

Steellgold commented:
> C'est déjà + ok, ça m'embête de prendre du temps sur un truc pas forcément permanent, mais oui tu peux.

@Bluzzi Ce n'est pas pertinent mais pour un visiteur lambda se serra + beau que répéter dans le même ordre constant

Steellgold committed:
feat(members-galery): add members shuffle functionality

This commit adds a new utility function to shuffle members in the
members-galery component. The shuffle function was moved to a new file
called members-shuffle.ts and a test file was created to test the
shuffleMembers function. The members-galery component now imports the
shuffleMembers function from the new file and uses it to shuffle the
members array before rendering.

Bluzzi commented on src/lib/utils/members-shuffle/members-shuffle.ts:
```diff
@@ -0,0 +1,12 @@
+import type { Member } from "#/lib/configs/member/member.type";
+
+export const shuffleMembers = (members: Member[]): Member[] => {
+  const shuffledMembers = [...members];
+
+  for (let i = shuffledMembers.length - 1; i > 0; i--) {
+    const j = Math.floor(Math.random() * (i + 1));
+    [shuffledMembers[i], shuffledMembers[j]] = [shuffledMembers[j], shuffledMembers[i]];
+  }
+
+  return shuffledMembers;
+};
```
La réutilisabilité ! Pourquoi se limiter à une array de `Member` quand on pourrait avoir une fonction permettant de shuffle n'importe quoi ?

Steellgold replied:
Oui effectivement j'y ai pas pensé directement, j'ai modifié et fait en sorte que l'on puisse mettre n'importe quel type de tableau dedans

Bluzzi commented on src/lib/utils/members-shuffle/members-shuffle.test.ts:
```diff
@@ -0,0 +1,13 @@
+import { describe, it, expect } from "vitest";
+import { shuffleMembers } from "./members-shuffle";
+import { members } from "#/lib/configs/member";
+
+describe("shuffleMembers", () => {
+  it("should return a shuffled array", () => {
+    expect(shuffleMembers(members)).not.toBe(members);
+  });
+
+  it("a empty array should return a empty array", () => {
+    expect(shuffleMembers([])).toEqual([]);
+  });
+});
```
Il faut aller plus loin, voici quelques tests à écrire :
- vérification de la non-mutabilité de la fonction
- vérification de la valeur de retour avec un mock de `Math.random` pour que ça sois prévisible 
- vérifier que l'array mélangé contient chaque élément de l'array précédente 
- vérifie que l'array a la même taille que l'array précédente 

Aussi, il ne faut pas utiliser `toBe()` mais `toEqual`, car `toBe` vérifie aussi l'authenticité de l'array, ici on veux juste vérifier leur contenu (donc dans ton actuel actuel le premier test ne pourra jamais fail).

N'utilise pas une de nos configurations mais directement une array (voir plusieurs) que tu créé dans le fichier de test, assure toi d'avoir au moins 8 valeurs dans l'array pour limiter un maximum les fails du test du à la mal chance d'un mélange qui retournerais la même array ahah.

8 valeurs ça fait 16 777 216 possibilités, ça me semble est suffisamment sur tout en restant pas trop lourd.

Steellgold committed:
refactor(members-galery.tsx): rename shuffleMembers to shuffle and move it to a new file
test(shuffle.test.ts): add tests for shuffle function
feat(shuffle.ts): add shuffle function to shuffle an array of any type

Steellgold committed:
test(shuffle.test.ts): change toEqual to not.toEqual for better test coverage
test(shuffle.test.ts): add test to check if shuffle function is immutable

Steellgold committed:
test(shuffle.test.ts): improve test descriptions and readability

Bluzzi commented on src/lib/utils/shuffle/shuffle.test.ts:
```diff
@@ -0,0 +1,41 @@
+import { describe, it, expect, vi } from "vitest";
+import { shuffle } from "./shuffle";
+
+describe("shuffle function", () => {
```
Parmis les tests que j'ai cité dans la review précédente, tu ne test pas :
- vérification de la non-mutabilité de la fonction
- vérifier que l'array mélangé contient chaque élément de l'array précédente (peut-être que si, mais j'aimerais un `it` bien nommé par test d'écrit précédemment)

Bluzzi replied:
Ok je viens de voir que c'est vérifie plus bas quand même. Malgré ça, split d'avantage les tests, même si le code est souvent explicite, l'objectif et de pouvoir fix une erreur juste en connaissant le message du test qui a fail.

Bluzzi commented on src/lib/utils/shuffle/shuffle.test.ts:
```diff
@@ -0,0 +1,41 @@
+  it("should return a shuffled array", () => {
+    const array = [1, 2, 3, 4, 5, 6, 7, 8];
+    const shuffled = shuffle(array);
+    expect(shuffled).not.toEqual(array);
+    expect(shuffled).toEqual(expect.arrayContaining(array));
+    expect(shuffled.length).toEqual(array.length);
```
Certains des tests suivant doivent être split pour avoir un nom bien conforme à ce qui est testé.

Bluzzi commented on src/lib/utils/shuffle/shuffle.test.ts:
```diff
@@ -0,0 +1,41 @@
+  });
+
+  it("should return an empty array when given an empty array", () => {
+    expect(shuffle([])).toEqual([]); // logika
```
Même si c'est drôle, restons pro :p
